apps/rehash.c

```diff
@@ -125,7 +125,7 @@ static int add_entry(enum Type type, unsigned int hash, const char *filename,
     static HENTRY nilhentry;
     BUCKET *bp;
     HENTRY *ep, *found = NULL;
-    unsigned int ndx = (type + hash) % OSSL_NELEM(hash_table);
+    unsigned int ndx = (type + (unsigned long)(hash)) % OSSL_NELEM(hash_table);
```

Comment:
This is going to leave the same problem on ILP32 or ILP64 systems. And yes, most systems are LLP64 (most Microsoft) or LP64 (most Linux), that's not everything.

But: since this is a hashing function, and since unsigned integer "overflow" doesn't happen in C (it's fully defined as a modulo reduction), is this really a problem?

Reply:
This can be a problem where the same directory with hash symlinks is used by both a 32 and 64 bit system, where the file can't be find if the hash is created with one, but looked up with the other.

Reply:
I do not think it matters here. This is just for duplicates check.

Reply:
What about casting to `u long long`? This would solve the problem with data models.
> is this really a problem?

It could be a problem:
```c
#include "limits.h"
#include "stdio.h"

#define OSSL_NELEM(x)    (sizeof(x)/sizeof((x)[0]))
enum Type { TYPE_CERT=0, TYPE_CRL=1 };

#define SOME_TYPE int
static SOME_TYPE *hash_table[257];

int main() {
	enum Type type = TYPE_CRL;
	unsigned int hash = UINT_MAX;
	unsigned int ndx = (type + hash) % (OSSL_NELEM(hash_table));
	// 0
	printf("%u\n", ndx);
	ndx = (type + (unsigned long long)(hash)) % (OSSL_NELEM(hash_table));
	// 1
	printf("%u\n", ndx);
	return 0;
}
```

Reply:
The cast would be to `uint64_t` not `unsigned long long int`.

Reply:
> I do not think it matters here. This is just for duplicates check.

I completely agree with [tm8](https://github.com/openssl/openssl/pull/23708#discussion_r1507331775). Such a construction does not seem to me to be a real problem.